db: add a new option StrictSync

lni opened the pull request:
StrictSync indicates that the DB should only make incoming records visible
once they are fully written and synchronized onto the underlying disk. This
is RocksDB's default behavior. If not set (the default), incoming records
are made visible as soon as they are applied into the MemTable. This makes
writes more performant at the risk of losing some visible records after a
DB restart, e.g. crash.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/816)
<!-- Reviewable:end -->

lni commented:
@jbowens Thanks for the input. I've addressed issues raised above. 

To handle writes with both Sync = false and Sync = true when StrictSync is true, the behavior of syncQueue in internal/record has been changed. It no longer just take care of those batches with Sync=true, it looks after all batches when StrictSync = true or write Sync = true. A new boolean flag is added to each slot to indicate whether a fsync() is required. This pretty much changed it to a completionQueue, as all StrictSync = true or write Sync = true batches will be put into this queue and eventually have their wg.Done() called.

lni commented:
@jbowens Thanks for the review. I will get that break statement fixed and add tests as recommended in the next couple of days.
